mautic/mautic
# [DPMMA-1709] Rollback of changes in the form in case of alias error

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [✅]
| New feature/enhancement? (use the a.x branch)      | [❌]
| Deprecations?                          | [❌]
| BC breaks? (use the c.x branch)        | [❌]
| Automated tests included?              | [✅] 
| Related user documentation PR URL      | mautic/mautic-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation#... 
| Issue(s) addressed                     | Fixes #... 



#### Description:
Attempting to create a new form or update an existing one with long, similar aliases for fields results in a 500 error. The form is partially saved in the database and the ID is not sent back in the API response.
```
mautic.CRITICAL: Uncaught PHP Exception Doctrine\DBAL\Schema\SchemaException: "The column 'name_long_enough_to_caus' on table 'form_results_26_test' already exists." at vendor\doctrine\dbal\lib\Doctrine\DBAL\Schema\SchemaException.php line 123 {"exception":"[object] (Doctrine\\DBAL\\Schema\\SchemaException(code: 40): The column 'name_long_enough_to_caus' on table 'form_results_26_test' already exists. at \\vendor\\doctrine\\dbal\\lib\\Doctrine\\DBAL\\Schema\\SchemaException.php:123)"}
```
This PR introduces alias verification before saving to the database. If an attempt to reuse the same alias after shortening it to 24 characters is detected, a 400 error should be returned along with the message:
"Another field is already using this alias: name_long_enough_to_caus. Please choose another or leave blank to have it autogenerated."
The changes will be rolled back from the database, so the updated form should retain the actions and fields that were saved in the previous, correct API call.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Using the API, try to create a new form or update an existing one with at least two fields with similar aliases that do not differ in the first 24 characters , e.g.
`name_long_enough_to cause_error` and `name_long_enough_to_cause_error2`
3. Attempting to save the form should rollback changes, return error with code 400 and a message "Another field is already using this alias: name_long_enough_to_caus. Please choose another or leave blank to have it autogenerated."
